src/coreclr/src/vm/zapsig.cpp

```diff
@@ -625,34 +626,41 @@ Module *ZapSig::DecodeModuleFromIndex(Module *fromModule,
     CONTRACTL_END;
 
     Assembly *pAssembly = NULL;
+    NativeImage *nativeImage = fromModule->GetCompositeNativeImage();
+    uint32_t assemblyRefMax = (nativeImage != NULL ? 0 : fromModule->GetAssemblyRefMax());
 
-    if (index == 0)
-    {
-        pAssembly = fromModule->GetAssembly();
-    }
-    else
+    if (index < assemblyRefMax)
     {
-        if (index < fromModule->GetAssemblyRefMax())
+        if (index == 0)
         {
-            pAssembly = fromModule->LoadAssembly(RidToToken(index, mdtAssemblyRef))->GetAssembly();
+            pAssembly = fromModule->GetAssembly();
         }
         else
         {
-            index -= fromModule->GetAssemblyRefMax();
+            pAssembly = fromModule->LoadAssembly(RidToToken(index, mdtAssemblyRef))->GetAssembly();
+        }
+    }
+    else
+    {
+        index -= assemblyRefMax;
 
-            pAssembly = fromModule->GetNativeMetadataAssemblyRefFromCache(index);
+        pAssembly = fromModule->GetNativeMetadataAssemblyRefFromCache(index);
 
-            if(pAssembly == NULL)
+        if(pAssembly == NULL)
+        {
+            if (nativeImage != NULL)
+            {
+                pAssembly = nativeImage->LoadComponentAssembly(index);
+            }
+            else
```

Comment:
Why add a function here, but not in DecodeModuleFromIndexIfLoaded?

Reply:
Ouch, thanks for pointing that out. I missed this second method when applying the module index refactoring suggested by JanV.

Reply:
Fixed in 11th commit.